inc/Engine/Media/Images/Frontend.php

```diff
@@ -0,0 +1,328 @@
+	 *
+	 * @return bool Valid to be parsed or not.
+	 */
+	private function can_specify_dimensions_external_images() {
+		/**
+		 * Enable/Disable specify image dimensions for external images.
+		 *
+		 * @since 3.8
+		 *
+		 * @param bool Specify image dimensions for external images or not.
+		 */
+		return ini_get( 'allow_url_fopen' ) && apply_filters( 'rocket_specify_image_dimensions_for_distant', true );
+	}
+
+	/**
+	 * Assign width and height attributes to the img tag.
+	 *
+	 * @param string $image IMG tag.
+	 * @param string $width_height Width/Height attributes in ready state like [height="100" width="100"].
+	 *
+	 * @return string IMG tag after adding attributes otherwise return the input img when error.
+	 */
+	private function assign_width_height( string $image, $width_height ) {
+		// Remove old width and height attributes if found.
+		$changed_image = preg_replace( '/(height|width)=[\'"](?:\S+)*[\'"]/i', '', $image );
+		$changed_image = preg_replace( '/<\s*img/i', '<img ' . $width_height, $changed_image );
+
+		if ( null === $changed_image ) {
+			return $image;
+		}
+
+		return $changed_image;
+	}
+
+	/**
+	 * Check if the image exists, internal or external image.
+	 *
+	 * @param string $image    Image Url for external and Image absolute path for internal.
+	 * @param bool   $external If this image is external or not.
+	 *
+	 * @return bool If image exists or not.
+	 */
+	private function image_exists( $image, $external = false ) {
+		if ( ! $external ) {
+			return $this->filesystem->exists( $image );
+		}
+
+		$file_headers = get_headers( $image );
+		if ( ! $file_headers ) {
+			return false;
+		}
+
+		return strpos( $file_headers[0], '404' ) === false;
+	}
+
+	/**
+	 * Check if we can specify image dimensions for all images.
+	 *
+	 * @return bool Can we or not.
+	 */
+	private function can_specify_dimensions_images() {
+		if ( ! $this->options->get( 'image_dimensions', false ) ) {
+			return false;
+		}
+
+		/**
+		 * Filter images dimensions attributes
+		 *
+		 * @since 2.2
+		 *
+		 * @param bool Do the job or not.
+		 */
+		return apply_filters( 'rocket_specify_image_dimensions', true );
+	}
+
+	/**
+	 * Check if we can specify image dimensions for one image.
+	 *
+	 * @param string $image Full img tag.
+	 *
+	 * @return false|string false if we can't specify for this image otherwise get img src attribute.
+	 */
+	private function can_specify_dimensions_one_image( $image ) {
+		// Don't touch lazy-load file (no conflict with Photon (Jetpack)).
+		if (
+			false !== strpos( $image, 'data-lazy-original' )
+			||
+			false !== strpos( $image, 'data-no-image-dimensions' )
+			||
+			! preg_match( '/\s+src\s*=\s*[\'"](?<url>[^\'"]+)/i', $image, $src_match )
+		) {
+			return false;
+		}
+
+		return $src_match['url'];
+
+	}
+
+	/**
+	 * Get Image sizes.
+	 *
+	 * @param string $image_url Image url to get sizes for.
+	 *
+	 * @return array|false Get image sizes otherwise false.
```

Comment:
```suggestion
	 * @return string|false Get image sizes otherwise false.
```

With the change, it's a string instead of an array, right?

Reply:
Yes exactly, Done